crates/kubelet/src/server.rs

```diff
@@ -111,79 +118,143 @@ where
+            format!(
+                "Unsupported method {} for resource '{}'.",
+                method,
+                req.uri().path()
+            ),
+        ),
+        Some(_) | None => return_code(
+            StatusCode::NOT_FOUND,
+            format!("Unknown resource '{}'.", req.uri().path()),
+        ),
+    }
+}
+
+/// Extract and validate namespace/pod/container resource path.
+/// On error return response to return to client.
+fn extract_container_path(
+    path: Vec<&str>,
+) -> Result<(String, String, String), anyhow::Result<Response<Body>>> {
+    match path[..] {
+        [] => Err(return_code(
+            StatusCode::BAD_REQUEST,
+            format!("Please specify a namespace."),
+        )),
+        [_] => Err(return_code(
+            StatusCode::BAD_REQUEST,
+            format!("Please specify a pod."),
+        )),
+        [_, _] => Err(return_code(
+            StatusCode::BAD_REQUEST,
+            format!("Please specify a container."),
+        )),
+        [s, _, _] if s == "" => Err(return_code(
+            StatusCode::BAD_REQUEST,
+            format!("Please specify a namespace."),
+        )),
+        [_, s, _] if s == "" => Err(return_code(
+            StatusCode::BAD_REQUEST,
+            format!("Please specify a pod."),
+        )),
+        [_, _, s] if s == "" => Err(return_code(
+            StatusCode::BAD_REQUEST,
+            format!("Please specify a container."),
+        )),
+        [namespace, pod, container] => Ok((
+            namespace.to_string(),
+            pod.to_string(),
+            container.to_string(),
+        )),
+        [_, _, _, ..] => {
+            let resource = path
+                .iter()
+                .rev()
+                .map(|s| *s)
+                .collect::<Vec<&str>>()
+                .join("/");
+            Err(return_code(
+                StatusCode::NOT_FOUND,
+                format!("Unknown resource '{}'.", resource),
+            ))
         }
-        (&Method::POST, [_, "exec", _, _, _]) => post_exec(&*provider, &req),
-        _ => Response::builder()
-            .status(StatusCode::NOT_FOUND)
-            .body(Body::from("Not Found"))
-            .unwrap(),
-    };
-
-    Ok(response)
+    }
 }
 
 /// Return a simple status message
-fn get_ping() -> Response<Body> {
-    Response::new(Body::from("this is the Krustlet HTTP server"))
+fn get_ping() -> anyhow::Result<Response<Body>> {
+    Ok(Response::new(Body::from(
+        "this is the Krustlet HTTP server",
+    )))
+}
+
+/// Return a HTTP code and message.
+fn return_code(code: StatusCode, body: String) -> anyhow::Result<Response<Body>> {
+    Ok(Response::builder().status(code).body(Body::from(body))?)
 }
 
 /// Get the logs from the running WASM module
 ///
 /// Implements the kubelet path /containerLogs/{namespace}/{pod}/{container}
 async fn get_container_logs<T: Provider + Sync>(
     provider: &T,
-    req: &Request<Body>,
+    _req: &Request<Body>,
```

Comment:
Looks like we don't need this any more?

Reply:
Probably, I want sure if we wanted API consistency with `post_exec`, `get_container_logs` and any future service functions.

Reply:
I think we are ok to remove, especially because this is a private function